### ISSUE TITLE
Client seems to be leaking connections after timeouts calling a non responsive memcached server

### ISSUE THREAD

**Issue by eranharel** [spotify/folsom#63]
We had a temporary faulty production memcached server which caused a few hundreds of request timeouts. When this happened the client did reconnect, but the old connections seem to remain open, and never got closed.

I have managed to recreate this behaviour using a stand-alone test.

I start 2 nodes like so:

```
memcached -m 64 -p 11213 -u memcache -l 127.0.0.1
memcached -m 64 -p 11214 -u memcache -l 127.0.0.1
```

I register them in consul so the DnsSrvResolver will pick them up. I omitted this code, cause it doesn't matter to this test.
I then execute the following main class:

```
public class ReconnetTest {

  private static final Logger log = LoggerFactory.getLogger(ReconnetTest.class);
  private static final String PAYLOAD = "012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789";

  public static void main(String[] args) throws ExecutionException, InterruptedException {
    final MemcacheClient<Serializable> client = MemcacheClientBuilder.newSerializableObjectClient()
      .withSRVRecord("memcached")
      .withSrvResolver(new HealthyMemcachedResolver())
      .withSRVRefreshPeriod(1000)
      .withRequestTimeoutMillis(100)
      .withSRVShutdownDelay(1000)
      .withBackoff(reconnectAttempt -> 2000)
      .connectBinary();
    ConnectFuture.connectFuture(client).get();

    final String key = "kkk";

    for (int i = 0; i < 100000000; i++) {
      try {
        client.set(key, i + "-" + PAYLOAD, 100000).get();
        client.get(key).get();
      } catch (final Exception e) {
        log.error(i + " " + e.getMessage());
      }
      Thread.sleep(1);
    }
  }

  private static class HealthyMemcachedResolver implements DnsSrvResolver {

    public static final ArrayList<LookupResult> NODES = Lists.newArrayList(LookupResult.create("localhost", 11213, 1, 1, 1000), LookupResult.create("localhost", 11214, 1, 1, 1000));

    @Override
    public List<LookupResult> resolve(String fqdn) {
      return NODES;
    }
  }
}
```

Then I move the server to the background for a few sec (ctrl+z), and return it to foreground.
I verify that the connections are still open:

```
~ $ netstat | grep 11214 | grep ESTABLISHED | wc -l
66
```

I think I have a fix for this - adding an `IdleStateHandler` to the pipeling, and forcing close() when idle channel. I will submit a PR if you verify this, and once I have a better test that can be executed in CI.
When I add my code, the idle connections get closed after the idle timeout.

Hopefully I'm not wasting anyone's time with a bogus configuration again ;)

We're using the **0.7.2** client.

**Comment by spkrka** [comment]
Thank you for the report! I don't have much time to look into in detail right now, but I'd like to understand more about why the connections aren't closed before adding bandaids (hopefully the underlying bug can be fixed instead).

Is this a minimal test? I don't quite understand why you need two memcached instances. I guess you moving the servers to the background will timeout requests to the memcached, so folsom should close and retry, and somewhere at this point one or more of the connections is not properly closed.

**Comment by eranharel** [comment]
Sorry for the late response, I wanted to have something useful to add to this report...

It turns out my initial test doesn't recreate the bug :(
I tested it with the previous folsom version by mistake. Once I upgraded the test doesn't recreate the bug.

How ever, in production I see a growing number of `com.spotify.folsom.client.DefaultRawMemcacheClient` instances when the connection count increases when I produce a class histogram.
I read the code several times and it seems to be correct, but I may mistaken, or maybe the `channel.close()` doesn't complete for some reason.

I took several heap dumps, and will try to analyse tomorrow once the download completes.
